File: uva/accordian.cpp
#include <bits/stdc++.h>
using namespace std;

struct card {
  char face;
  char suit;

  card(char f, char s){
    face = f;
    suit = s;
  }

  bool operator == (card c){
    return (face == c.face || suit == c.suit);
  }
};

void display(list<stack<card>> &l){
  int sz = l.size();
  if (sz == 1){
    cout << sz << " pile remaining: ";
  } else {
    cout << sz << " piles remaining: ";
  }

  for (auto iter = l.begin(); iter != l.end(); ++iter){
    cout << iter->size() << " \n"[--sz==0];
  }
}

void move(stack<card> &s1, stack<card> &s2){
  //cout << s1.top().face << s1.top().suit << " over " << s2.top().face << s2.top().suit << " >>> ";
  s2.emplace(s1.top().face, s1.top().suit);
  s1.pop();
  //cout << s2.top().face << s2.top().suit << '\n';
}

int main(){
  char f,s;
  while (true){
    list<stack<card>> l;
    stack<card> *stk;

    for (int i=0; i<52; ++i){
      cin >> f >> s;
      if (f == '#') return 0;
      stk = new stack<card>;
      stk->emplace(f,s);
      l.push_back(*stk);
    }

    //display(l);

    int pos = 0;
    list<stack<card>>::iterator iter1 = l.begin(), iter2;
    while (iter1 != l.end()){
      //display(l);
      //cout << pos << ' ' << iter1->top().face << iter1->top().suit << '\n';
      //cout << pos << ' ' << l.size() << '\n';
      if (pos >= 3){
        iter2 = prev(iter1,3);
        if (iter1->top() == iter2->top()){
          move(*iter1, *iter2);
          //cout << s2.top().face << s2.top().suit << '\n';
          if (iter1->empty()){
            iter1 = l.erase(iter1);
          }
          pos-=3;
          iter1 = prev(iter1,3);
          //cout << iter1->top().face << iter1->top().suit << '\n';
          continue;
        }
      }

      if (pos >= 1){
        iter2 = prev(iter1,1);
        if (iter1->top() == iter2->top()){
          move(*iter1, *iter2);
          //cout << s2.top().face << s2.top().suit << '\n';
          if (iter1->empty()){
            iter1 = l.erase(iter1);
          }
          --pos;
          --iter1;
          //cout << iter1->top().face << iter1->top().suit << '\n';
          continue;
        }
      }

      ++pos;
      ++iter1;
    }

    display(l);
  }

  return 0;
}

File: uva/accordian.py
class Card():
    def __init__(self, f, s):
        self.face, self.suit = f, s

    def __repr__(self):
        return self.face + self.suit

def match(s1, s2):
    return s1[-1].face == s2[-1].face or s1[-1].suit == s2[-1].suit

def move(s1, s2):
    s2.append(s1.pop())

def display(cards):
    l = len(cards)
    p = " ".join(str(len(s)) for s in cards)
    if (l > 1):
        print("{} piles remaining: {}".format(l,p))
    else:
        print("{} pile remaining: {}".format(l,p))

if __name__ == "__main__":
    while True:
        cards = []
        for i in range(2):
            for c in input().split():
                if c == "#": exit()
                cards.append([Card(c[0],c[1])])
        i = 0
        while (i < len(cards)):
            if (i >= 3):
                if match(cards[i], cards[i-3]):
                    move(cards[i], cards[i-3])
                    if len(cards[i]) == 0:
                        del cards[i]
                    i-=3
                    continue
            if (i >= 1):
                if match(cards[i], cards[i-1]):
                    move(cards[i], cards[i-1])
                    if len(cards[i]) == 0:
                        del cards[i]
                    i-=1
                    continue
            i+=1

        display(cards)

"""
    algorithm

    for each case
        store inputs as stacks of 1 card in a list
        set card stack position cur = 0
        while cur <= last stack position
            if cur-3 >= 0, and cur and cur-3 top cards match
                move top card at cur to cur-3
                if cur stack is empty, remove from list
                set cur = cur-3
                continue loop
            if cur-1 >= 0, and cur and cur-1 top cards match
                move top card at cur to cur-1
                if cur stack is empty, remove from list
                set cur = cur-1
                continue loop
            no matches, set cur = cur+1, continue loop
        print the final arrangement of cards

"""
